etc/gen_gen/src/bin/gen_gen.rs

```diff
@@ -0,0 +1,324 @@
+	let keychain = ExtKeychain::from_random_seed().unwrap();
+	let key_id = ExtKeychain::derive_key_id(0, 1, 0, 0, 0);
+	let reward = core::libtx::reward::output(&keychain, &key_id, 0, 0).unwrap();
+	gen = gen.with_reward(reward.0, reward.1);
+
+	{
+		// setup a tmp chain to set block header roots
+		core::global::set_mining_mode(core::global::ChainTypes::AutomatedTesting);
+		let tmp_chain = setup_chain(".grin.tmp", core::pow::mine_genesis_block().unwrap());
+		tmp_chain.set_txhashset_roots(&mut gen).unwrap();
+	}
+
+	// mine a Cuckaroo29 block
+	core::global::set_mining_mode(core::global::ChainTypes::Mainnet);
+	let plugin_lib = cuckoo::PluginLibrary::new(PLUGIN_PATH).unwrap();
+	let mut params = plugin_lib.get_default_params();
+	params.mutate_nonce = false;
+	let solver_ctx = plugin_lib.create_solver_ctx(&mut params);
+
+	let mut solver_sols = plugin::SolverSolutions::default();
+	let mut solver_stats = plugin::SolverStats::default();
+	let mut nonce = 0;
+	while solver_sols.num_sols == 0 {
+		solver_sols = plugin::SolverSolutions::default();
+		gen.header.pow.nonce = nonce;
+		let _ = plugin_lib.run_solver(
+			solver_ctx,
+			gen.header.pre_pow(),
+			nonce,
+			1,
+			&mut solver_sols,
+			&mut solver_stats,
+		);
+		if solver_stats.has_errored {
+			println!(
+				"Plugin {} has errored, device: {}. Reason: {}",
+				solver_stats.get_plugin_name(),
+				solver_stats.get_device_name(),
+				solver_stats.get_error_reason(),
+			);
+			return;
+		}
+
+		nonce += 1;
+	}
+
+	// Set the PoW solution and make sure the block is mostly valid
+	gen.header.pow.proof.nonces = solver_sols.sols[0].to_u64s();
+	assert!(gen.header.pow.is_secondary(), "Not a secondary header");
+	println!("Built genesis:\n{:?}", gen);
+	core::pow::verify_size(&gen.header).unwrap();
+	gen.validate(
+		&BlindingFactor::zero(),
+		Arc::new(util::RwLock::new(LruVerifierCache::new())),
+	)
+	.unwrap();
+
+	println!("\nFinal genesis cyclehash: {}", gen.hash().to_hex());
+	let gen_bin = core::ser::ser_vec(&gen).unwrap();
+	println!("Final genesis full hash: {}\n", gen_bin.hash().to_hex());
+
+	update_genesis_rs(&gen);
+	println!("genesis.rs has been updated, check it and run mainnet_genesis_hash test");
+	println!("also check bitcoin block {} hasn't been orphaned.", h1);
+	println!("press c+enter to proceed.");
+	let mut input = String::new();
+	io::stdin().read_line(&mut input).unwrap();
+	if input != "c\n" {
+		return;
+	}
+
+	// Commit genesis block info in git and tag
+	process::Command::new("git")
+		.args(&["commit", "-am", "Minor: finalized genesis block"])
+		.status()
+		.expect("git commit failed");
+	process::Command::new("git")
+		.args(&["tag", "-a", "v1.0", "-m", "Mainnet release"])
+		.status()
+		.expect("git tag failed");
+	process::Command::new("git")
+		.args(&["push", "origin", "v1.0"])
+		.status()
+		.expect("git tag push failed");
+	println!("All done!");
+}
+
+fn update_genesis_rs(gen: &core::core::Block) {
```

Comment:
Took me a while to get my head aorund what this is doing. We're actually going into the `genesis.rs` file and replacing strings? 
Feels kind of fragile with the looking for `REPLACE` logic.

Maybe worth considering using a similar approach to simply write out a `genesis.toml` file and then `genesis.rs` can read it in?
That way its a lot more explicit what params were "replaced" etc.

Reply:
I did hesitate in the approach with some external data file (FWIW that's how Ethereum did it). It does feel cleaner and replacing in `genesis.rs` is a tad weird. But in the end I still opted for replacement in `genesis.rs`, even though as you mention it feels more fragile, because distribution is straightforward (git commit+compile), I can rely on the Rust compiler to catch some mistakes and the result is simpler (the genesis definition you'd expect in `genesis.rs`). You'll also see a clean diff on the commit, so saying what was replaced is actually very easy.

With an external file, distribution is more open. Is it downloaded? Or loaded from the repo? It's also another way to define genesis, when so far they've all been defined programmatically. Comparison between the 2 definitions is also harder. And finally, it would eventually be simpler to just update the genesis definition in `genesis.rs` after a few days. So why not dot that in the first place?